File: services/inventory-service/internal/infra/consumer.go
package infra

import (
	"context"
	"fmt"
	"log/slog"

	amqp "github.com/rabbitmq/amqp091-go"
	"google.golang.org/protobuf/encoding/protojson"

	eventsv1 "mirea-crm/gen/go/mirea/events/v1"
)

// Handler обрабатывает одно событие. Возвращённая ошибка означает, что
// сообщение уйдёт в dead-letter: повторять его бессмысленно или опасно.
type Handler func(ctx context.Context, envelope *eventsv1.EventEnvelope) error

type Consumer struct {
	conn    *amqp.Connection
	channel *amqp.Channel
	queue   string
	routes  map[string]Handler
}

func NewConsumer(url, queue string) (*Consumer, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, fmt.Errorf("подключение к RabbitMQ: %w", err)
	}

	channel, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("канал RabbitMQ: %w", err)
	}
	// Без ограничения брокер вывалит в потребителя всю очередь разом,
	// и при падении процесса вся пачка вернётся необработанной.
	if err := channel.Qos(16, 0, false); err != nil {
		_ = conn.Close()
		return nil, fmt.Errorf("prefetch: %w", err)
	}

	return &Consumer{conn: conn, channel: channel, queue: queue, routes: map[string]Handler{}}, nil
}

func (c *Consumer) Handle(routingKey string, handler Handler) {
	c.routes[routingKey] = handler
}

func (c *Consumer) Close() {
	if c.channel != nil {
		_ = c.channel.Close()
	}
	if c.conn != nil {
		_ = c.conn.Close()
	}
}

// Run читает очередь до отмены контекста. Очередь и биндинги объявлены
// декларативно в deploy/rabbitmq/definitions.json, потребитель их не создаёт.
func (c *Consumer) Run(ctx context.Context) error {
	deliveries, err := c.channel.ConsumeWithContext(ctx, c.queue, "", false, false, false, false, nil)
	if err != nil {
		return fmt.Errorf("подписка на %s: %w", c.queue, err)
	}
	slog.Info("слушаем очередь", "queue", c.queue)

	for {
		select {
		case <-ctx.Done():
			return nil
		case delivery, ok := <-deliveries:
			if !ok {
				return fmt.Errorf("канал доставки закрыт")
			}
			c.dispatch(ctx, delivery)
		}
	}
}

func (c *Consumer) dispatch(ctx context.Context, delivery amqp.Delivery) {
	var envelope eventsv1.EventEnvelope
	if err := protojson.Unmarshal(delivery.Body, &envelope); err != nil {
		slog.Error("не разобрали событие", "error", err, "message_id", delivery.MessageId)
		_ = delivery.Nack(false, false)
		return
	}

	// Трасса продолжается: контекст пришёл вместе с событием.
	ctx = WithTraceparent(ctx, ParseTraceparent(envelope.GetTraceparent()))
	log := slog.With(
		"routing_key", envelope.GetRoutingKey(),
		"event_id", envelope.GetEventId(),
		"trace_id", TraceID(Traceparent(ctx)),
	)

	handler, ok := c.routes[envelope.GetRoutingKey()]
	if !ok {
		// На ключ никто не подписан — подтверждаем, иначе очередь встанет.
		log.Warn("обработчик не найден")
		_ = delivery.Ack(false)
		return
	}

	if err := handler(ctx, &envelope); err != nil {
		log.Error("обработка не удалась", "error", err)
		_ = delivery.Nack(false, false)
		return
	}

	log.Debug("обработано")
	_ = delivery.Ack(false)
}

File: services/billing-service/app/infra/consumer.py
"""Каркас потребителя RabbitMQ.

Очередь и биндинги объявлены декларативно в deploy/rabbitmq/definitions.json —
потребитель их не создаёт, только подписывается.
"""

import asyncio
import logging
from collections.abc import Awaitable, Callable

import aio_pika
from google.protobuf.json_format import Parse, ParseError
from mirea.events.v1 import events_pb2

from app.infra import tracing

log = logging.getLogger(__name__)

Handler = Callable[[events_pb2.EventEnvelope], Awaitable[None]]

# Без ограничения брокер вывалит в потребителя всю очередь разом, и при
# падении процесса вся пачка вернётся необработанной.
PREFETCH = 16


class Consumer:
    def __init__(self, amqp_url: str, queue: str) -> None:
        self._url = amqp_url
        self._queue_name = queue
        self._routes: dict[str, Handler] = {}
        self._fallback: Handler | None = None
        self._connection: aio_pika.abc.AbstractRobustConnection | None = None
        self._task: asyncio.Task | None = None

    def handle(self, routing_key: str, handler: Handler) -> None:
        self._routes[routing_key] = handler

    def handle_all(self, handler: Handler) -> None:
        """Один обработчик на весь поток — для подписчиков с биндингом `#`."""
        self._fallback = handler

    async def start(self) -> None:
        self._connection = await aio_pika.connect_robust(self._url)
        channel = await self._connection.channel()
        await channel.set_qos(prefetch_count=PREFETCH)

        queue = await channel.get_queue(self._queue_name)
        self._task = asyncio.create_task(queue.consume(self._on_message, no_ack=False))
        log.info("слушаем очередь %s", self._queue_name)

    async def stop(self) -> None:
        if self._task is not None:
            self._task.cancel()
        if self._connection is not None:
            await self._connection.close()

    async def _on_message(self, message: aio_pika.abc.AbstractIncomingMessage) -> None:
        try:
            envelope = Parse(message.body.decode(), events_pb2.EventEnvelope())
        except (ParseError, UnicodeDecodeError):
            log.exception("не разобрали событие, message_id=%s", message.message_id)
            await message.reject(requeue=False)
            return

        # Трасса продолжается: контекст пришёл вместе с событием.
        tracing.set_current(tracing.parse(envelope.traceparent))

        handler = self._routes.get(envelope.routing_key, self._fallback)
        if handler is None:
            # На ключ никто не подписан — подтверждаем, иначе очередь встанет.
            log.warning("обработчик не найден: %s", envelope.routing_key)
            await message.ack()
            return

        try:
            await handler(envelope)
        except Exception:
            log.exception(
                "обработка не удалась: %s, event_id=%s, trace_id=%s",
                envelope.routing_key, envelope.event_id, tracing.trace_id(),
            )
            # Повторять бессмысленно или опасно — в dead-letter.
            await message.reject(requeue=False)
            return

        await message.ack()
